lnwallet/btcwallet/btcwallet.go

```diff
@@ -1199,33 +1200,98 @@ func (b *BtcWallet) ListUnspentWitness(minConfs, maxConfs int32,
 // already published to the network (either in the mempool or chain) no error
 // will be returned.
 func (b *BtcWallet) PublishTransaction(tx *wire.MsgTx, label string) error {
-	if err := b.wallet.PublishTransaction(tx, label); err != nil {
+	// handleErr is a helper closure that handles the error from
+	// PublishTransaction and TestMempoolAccept.
+	handleErr := func(err error) error {
+		var (
+			doubleSpendErr      *base.ErrDoubleSpend
+			replacementErr      *base.ErrReplacement
+			mempoolFeeErr       *base.ErrMempoolFee
+			inMempoolErr        *base.ErrInMempool
+			alreadyConfirmedErr *base.ErrAlreadyConfirmed
+		)
+
 		// If we failed to publish the transaction, check whether we
 		// got an error of known type.
-		switch err.(type) {
+		switch {
 		// If the wallet reports a double spend, convert it to our
 		// internal ErrDoubleSpend and return.
-		case *base.ErrDoubleSpend:
+		case errors.As(err, &doubleSpendErr):
 			return lnwallet.ErrDoubleSpend
 
 		// If the wallet reports a replacement error, return
 		// ErrDoubleSpend, as we currently are never attempting to
 		// replace transactions.
-		case *base.ErrReplacement:
+		case errors.As(err, &replacementErr):
 			return lnwallet.ErrDoubleSpend
 
-		// If the wallet reports that fee requirements for accepting the
-		// tx into mempool are not met, convert it to our internal
+		// If the wallet reports that fee requirements for accepting
+		// the tx into mempool are not met, convert it to our internal
 		// ErrMempoolFee and return.
-		case *base.ErrMempoolFee:
+		case errors.As(err, &mempoolFeeErr):
 			return fmt.Errorf("%w: %v", lnwallet.ErrMempoolFee,
 				err.Error())
 
+		//nolint:lll
+		// These two errors are ignored inside `PublishTransaction`:
+		// https://github.com/btcsuite/btcwallet/blob/master/wallet/wallet.go#L3763
+		// To keep our current behavior, we need to ignore the same
+		// errors returned from TestMempoolAccept.
+		//
+		// TODO(yy): since `LightningWallet.PublishTransaction` always
+		// publish the same tx twice, we'd always get ErrInMempool. We
+		// should instead create a new rebroadcaster that monitors the
```

Comment:
Yeah I also hacked around it a bit, so have it eat that first error so it could continue publishing instead of bailing out as soon as things started up.

Reply:
yeah plan to implement a mempool-based rebroadcaster - think the current rebroadcaster suits neutrino well as neutrino has no mempool. For full node we could instead make use of the info found in mempool, and only republish a tx when it drops out the mempool.